### DISCUSSION TITLE
wire:navigate ignores the domain it is about to redirect to

### DISCUSSION THREAD

**Discussion by CapitaineToinon** [livewire/livewire#8786] [7. Bugs (archived)]
### Livewire version

3.5.6

### Laravel version

11.21.0

### Which PHP version are you using?

PHP 8.3

### Steps To Reproduce

Create a simple link in a blade component to an external website with wire:navigate:

```html
<a wire:navigate href="https://www.google.ch">Google</a>
```

The user clicking the link will quietly be redirected to the homepage in this instance. 

### Repository or Wirebox URL

https://github.com/CapitaineToinon/livewire-navigate-external

### Additional information

My use case is that I'm adding wire:navigate to all links in generated markdown content from a database. In my opinion, wire:navigate should either:

1. Fallback to a regular navigation whenever anything goes wrong or when the link is external
2. Throw an error at runtime whenever something fails

Right now, the link will quietly replace the page with the content from `/` with no error shown either to the user, to the Javascript console or the php console.

### Please confirm (incomplete submissions will not be addressed)

- [X] I have provided easy and step-by-step instructions to reproduce the bug.
- [X] I have provided code samples as text and NOT images.
- [X] I understand my bug report will be removed if I haven't met the criteria above.

**Comment by EdgarsJoja** [comment]
I agree. IMO regular redirect would be the best fallback solution.
Had the same issue and created helper function to add `wire:navigate` when necessary.
```php
if (!function_exists('wire_navigate')) {
    function wire_navigate(string $url): string
    {
        return !str_starts_with($url, 'http') ? 'wire:navigate' : '';
    }
}
```
But it would be better if this is handled by Livewire

**Comment by CapitaineToinon** [comment]
I've done some digging and it seems that I am not intended to fetch from other domains and that the navigation is supposed to error. However the code doesn't seems to make any sense to me on why it would actually error and in practice, that never happens for me:

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/js/plugins/navigate/history.js#L132-L147

The root problem seems to be the use of the `getUriStringFromUrlObject` which by definition, omits the domain you are about to navigate to.

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/js/plugins/navigate/links.js#L76-L78

So you can see I boosted a link with `https://www.google.ch`:

![image](https://github.com/user-attachments/assets/5c301475-ad21-4f46-acf9-1f3ce5195a08)

But now I'm fetching `/`:

![image](https://github.com/user-attachments/assets/66d42a25-3ec7-4ba3-baf5-113fec9cf6b6)

So I was wrong to say that it fails silently by redirecting to `/`. It just boosts the links by removing the domain, which is even worse. So if I try to boost to `https://www.google.ch/about-us` and I have an `/about-us` page on my side, it will redirect to `/about-us` just fine. What the hell?

So yeah this is all very confusing. Also I checked the Livewire documentation and it makes no mention of boosting external links. Looking at the code, it seems to have no guardrails and just assume that you're about to boost to an internal link even when that is not the case.

**Reply by CapitaineToinon** [reply]
Those two links do the exact same thing:

```html
<li>
    <a wire:navigate href="/about-us">
        About Us</a>
</li>
<li>
    <a wire:navigate href="https://www.google.ch/about-us">
        Google</a>
</li>
```

**Comment by CapitaineToinon** [comment]
For now, it seems like I can do the check myself:

```js
document.addEventListener('livewire:navigate', (event) => {
	/** @type{URL} */
	const to = event.detail.url
	const here = new URL(window.location.href)

	if (to.origin !== here.origin) {
                event.preventDefault()
                window.location.assign(to);
	}
})
```

This is the behavior I personally believe makes the most sense. I still think Livewire should fix this problem though.

**Reply by CapitaineToinon** [reply]
> For now, it seems like I can do the check myself:
> 
> ```js
> document.addEventListener('livewire:navigate', (event) => {
> 	/** @type{URL} */
> 	const to = event.detail.url
> 	const here = new URL(window.location.href)
> 
> 	if (to.origin !== here.origin) {
>                 event.preventDefault()
>                 window.location.assign(to);
> 	}
> })
> ```
> 
> This is the behavior I personally believe makes the most sense. I still think Livewire should fix this problem though.

So turns out this is not great since it prevents opening in new tabs and such. This problem is such a pain in the butt to deal with. I suggest people to be very careful when they use wire:navigate. I made a custom link blade component that checks if the link is internal or not and conditionally apply the directive but yeah. I really wish this could be fixed.

```php
<?php

namespace App\View\Components;

use Closure;
use Illuminate\Contracts\View\View;
use Illuminate\View\Component;
use Str;

class Link extends Component
{
    private static function hostMatches(string $host, array $compareTo): bool
    {
        foreach ($compareTo as $c) {
            if (\strpos($c, '/') === 0) {
                if (\preg_match($c, $host)) {
                    return true;
                }
            } elseif ($c === $host) {
                return true;
            }
        }

        return false;
    }

    private static function isExternal(?string $href): bool
    {
        if ($href === null) {
            return false;
        }

        $host = \parse_url($href, PHP_URL_HOST);

        if (! \is_string($host)) {
            return false;
        }

        if (self::hostMatches($host, ['localhost'])) {
            return false;
        }

        return true;
    }

    /**
     * Create a new component instance.
     */
    public function __construct(
        public ?string $href,
        public ?string $title,
        public ?bool $externalIcon = false) {}

    /**
     * Get the view / contents that represent the component.
     */
    public function render(): View|Closure|string
    {
        $attributes = [
            'rel' => null,
            'target' => null,
            'navigate' => ! Str::startsWith($this->href, '#'),
        ];

        $is_external = self::isExternal($this->href);
        $show_icon = $is_external && $this->externalIcon;

        if ($is_external) {
            $attributes = [
                'rel' => 'noopener noreferrer',
                'target' => '_blank',
                'navigate' => false,
            ];
        }

        return view('components.link', [
            'show_icon' => $show_icon,
            ...$attributes,
        ]);
    }
}
```

```blade
<a
    href="{{ $href }}"
    title="{{ $title }}"
    @if($rel) rel="{{ $rel }}" @endif
    @if($target) target="{{ $target }}" @endif
    @if($navigate) wire:navigate @endif
    {{ $attributes }}
>{{ $slot }}@if ($show_icon)<flux:icon.arrow-top-right-on-square class="size-4 ml-1 inline-block mb-1"></flux:icon>@endif
</a>
```

**Comment by CapitaineToinon** [comment]
Okay I understand what's going on with the try catch, `history.pushState` is supposed to throw when redirecting to another domain. However the problem is that in `fetchHtml`, like I said the domain is stripped by `getUriStringFromUrlObject`

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/js/plugins/navigate/fetch.js#L4-L10

Then that uri is sent to `performFetch` which builds back a URL Object but using the response from `fetch`:

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/js/plugins/navigate/fetch.js#L26-L28

And finally, that `finalDestination` is then used for navigation, which will always be on the correct domain:

https://github.com/livewire/livewire/blob/597a2808d8d3001cc3ed5ce89a6ebab00f83b80f/js/plugins/navigate/index.js#L89-L93

And so the try catch never fires.
